File: src/solutions/common/bizz/cityapp.py
# -*- coding: utf-8 -*-
# Copyright 2020 Green Valley Belgium NV
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# @@license_version:1.7@@

import json
import logging
import urllib
from datetime import datetime
from types import NoneType

from google.appengine.api import urlfetch

from mcfw.rpc import returns, arguments
from rogerthat.rpc import users
from solutions.common.dal.cityapp import get_uitdatabank_settings
from solutions.common.models.cityapp import UitdatabankSettings


@returns(NoneType)
@arguments(service_user=users.User, version=unicode, params=dict)
def save_uitdatabank_settings(service_user, version, params):
    settings = get_uitdatabank_settings(service_user)
    settings.version = version
    settings.params = params
    settings.put()


@returns(tuple)
@arguments(settings=UitdatabankSettings, page=(int, long), pagelength=(int, long), changed_since=(int, long, NoneType))
def get_uitdatabank_events(settings, page, pagelength, changed_since=None):
    if settings.version == UitdatabankSettings.VERSION_3:
        return _get_uitdatabank_events_v3(settings, page, pagelength, changed_since)

    return False, 0, "Incorrect API version"


@returns(tuple)
@arguments(settings=UitdatabankSettings, page=(int, long), pagelength=(int, long), changed_since=(int, long, NoneType))
def _get_uitdatabank_events_v3(settings, page, pagelength, changed_since=None):
    key = settings.params.get('key')
    postal_codes = settings.params.get('postal_codes') or []

    if not key or not postal_codes:
        return False, 0, "Not all fields are provided"

    q_array = []
    for postal_code in postal_codes:
        q_array.append('address.nl.postalCode:%s' % postal_code)

    if settings.params.get('test'):
        url = "https://search-test.uitdatabank.be/events?"
    else:
        url = "https://search.uitdatabank.be/events?"
    values = {'apiKey': key,
              'start': (page - 1) * pagelength,
              'limit': pagelength,
              'calendarType': 'single,multiple,periodic',
              'q': ' OR '.join(q_array)}
    if changed_since:
        values['modifiedFrom'] = datetime.utcfromtimestamp(changed_since).isoformat() + '+00:00'
    data = urllib.urlencode(values)
    logging.debug(url + data)
    try:
        result = urlfetch.fetch(url + data, deadline=30)
    except urlfetch.DownloadError as e:
        logging.debug('Caught %s. Retrying....', e.__class__.__name__)
        result = urlfetch.fetch(url + data, deadline=30)
    if result.status_code != 200:
        logging.info("_get_uitdatabank_events_v3 failed with status_code %s and content:\n%s" %
                     (result.status_code, result.content))
        return False, 0, "Make sure your credentials are correct."
    r = json.loads(result.content)

    if r['totalItems'] == 0:
        if changed_since:
            return True, 0, []
        return False, 0, "0 upcoming events. Make sure your postal codes are correct."

    if r['totalItems'] > 10000:
        logging.error('Result has more then 10.000 items')

    return True, r['totalItems'], [member['@id'] for member in r['member']]

File: src/static/js/solutions/cityapp.js
/*
 * Copyright 2020 Green Valley Belgium NV
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @@license_version:1.7@@
 */
$(function() {
    'use strict';

    var saveUitdatabankSettings = function() {
        var allOK = true;
        var version = "3";
        var params = {};

        if ($('#section_agenda .sln-uit-events-3-key input').val() === "") {
            $('#section_agenda .sln-uit-events-3-key input').addClass("error");
            allOK = false;
        }
        if ($('#section_agenda .sln-uit-events-3-postal-codes').data('postal_codes') === null) {
            $('#section_agenda .sln-uit-events-3-postal-codes').addClass("error");
            allOK = false;
        } else {
            $('#section_agenda .sln-uit-events-3-postal-codes').removeClass("error");
        }

        params = {
            'key': $('#section_agenda .sln-uit-events-3-key input').val(),
            'postal_codes': $('#section_agenda .sln-uit-events-3-postal-codes').data('postal_codes')
        };

        if (allOK) {
            var data = JSON.stringify({
                version: version,
                params: params,
            });
            sln.call({
                url: "/common/cityapp/uitdatabank/settings",
                type: "POST",
                data: {
                    data: data
                },
                success: function(data) {
                    if (!data.success) {
                        sln.alert(data.errormsg, null, CommonTranslations.ERROR);
                    } else {
                    	checkUitdatabankSettings();
                    }
                },
                error: sln.showAjaxError
            });
        }
    };

    var checkUitdatabankSettings = function() {
    	sln.call({
            url: "/common/cityapp/uitdatabank/check",
            type: "GET",
            success: function(data) {
                $('#uitdatabankStatus').toggleClass('alert-danger', !data.enabled).toggleClass('alert-success', data.enabled);
                var text = CommonTranslations.STATUS_ENABLED;
                if (data.enabled) {
                    text = CommonTranslations.STATUS_ENABLED + " (" + data.count + " events found)";
                } else {
                	text = CommonTranslations.STATUS_DISABLED + (data.errormsg === undefined ? "" : (": " + data.errormsg));
                }
                $('#uitdatabankStatusText').text(text);
            },
            error: sln.showAjaxError
        });
    };

    var loadUitdatabankSettings = function() {
        sln.call({
            url: "/common/cityapp/uitdatabank/settings",
            type: "GET",
            success: function(data) {
            	checkUitdatabankSettings();

                $("#section_agenda .sln-uit-events-3").hide();

              	var postal_codes = data.params ? data.params.postal_codes : [];
                $("#section_agenda .sln-uit-events-3").show();
                $("#section_agenda .sln-uit-events-3-key input").val(data.params ? data.params.key : "");
                $("#section_agenda .sln-uit-events-3-postal-codes").data('postal_codes', postal_codes);
                renderPostalCodes(postal_codes);
            },
            error: sln.showAjaxError
        });
    };

    var deletePostalCode = function() {
        var region = $(this).attr('region');
        var htmlElement = $('.sln-uit-events-3-postal-codes');
        var index = htmlElement.data('postal_codes').indexOf(region);
        if (index >= 0) {
            htmlElement.data('postal_codes').splice(index, 1);
            renderPostalCodes(htmlElement.data('postal_codes'));
            saveUitdatabankSettings();
        }
    };

    var renderPostalCodes = function(postalCodes) {
        var htmlElement = $('.sln-uit-events-3-postal-codes table tbody');
        htmlElement.empty();
        var html = $.tmpl(templates.events_uitcalendar_settings, {
            regions: postalCodes
        });
        htmlElement.append(html);
        htmlElement.find('button[action="deleteRegion"]').click(deletePostalCode);
    };

    $(".sln-set-events-uit-settings").show();

    sln.configureDelayedInput($('.sln-uit-events-3-key input'), saveUitdatabankSettings);
    $('#sln-uit-events-3-add-postal-code').click(function() {
        sln.input(function(value) {
            if (!value.trim())
                return false;

            var elem = $('.sln-uit-events-3-postal-codes');
            elem.data('postal_codes').push(value);
            renderPostalCodes(elem.data('postal_codes'));
            saveUitdatabankSettings();

        }, CommonTranslations.ADD, CommonTranslations.SAVE, CommonTranslations.ENTER_DOT_DOT_DOT, null, null, null);
    });

    loadUitdatabankSettings();
});
